src/analyser.cpp

```diff
@@ -0,0 +1,1509 @@
+            equation->addOdeVariable(analyserVariable(variable));
+        } else if (!(node->parent()->isMathmlElement("bvar")
+                     && node->parent()->parent()->firstChild()->isMathmlElement("diff"))) {
+            equation->addVariable(analyserVariable(variable));
+        }
+
+        // Add the variable to our AST.
+
+        ast->mPimpl->populate(AnalyserEquationAst::Type::CI, variable, astParent);
+    } else if (node->isMathmlElement("cn")) {
+        if (mathmlChildCount(node) == 1) {
+            // We are dealing with an e-notation based CN value.
+
+            ast->mPimpl->populate(AnalyserEquationAst::Type::CN, node->firstChild()->convertToString() + "e" + node->firstChild()->next()->next()->convertToString(), astParent);
+        } else {
+            ast->mPimpl->populate(AnalyserEquationAst::Type::CN, node->firstChild()->convertToString(), astParent);
+        }
+
+        // Qualifier elements.
+
+    } else if (node->isMathmlElement("degree")) {
+        ast->mPimpl->populate(AnalyserEquationAst::Type::DEGREE, astParent);
+
+        processNode(mathmlChildNode(node, 0), ast->mPimpl->mLeftChild, ast, component, equation);
+    } else if (node->isMathmlElement("logbase")) {
+        ast->mPimpl->populate(AnalyserEquationAst::Type::LOGBASE, astParent);
+
+        processNode(mathmlChildNode(node, 0), ast->mPimpl->mLeftChild, ast, component, equation);
+    } else if (node->isMathmlElement("bvar")) {
+        ast->mPimpl->populate(AnalyserEquationAst::Type::BVAR, astParent);
+
+        processNode(mathmlChildNode(node, 0), ast->mPimpl->mLeftChild, ast, component, equation);
+
+        auto rightNode = mathmlChildNode(node, 1);
+
+        if (rightNode != nullptr) {
+            processNode(rightNode, ast->mPimpl->mRightChild, ast, component, equation);
+        }
+
+        // Constants.
+
+    } else if (node->isMathmlElement("true")) {
+        ast->mPimpl->populate(AnalyserEquationAst::Type::TRUE, astParent);
+    } else if (node->isMathmlElement("false")) {
+        ast->mPimpl->populate(AnalyserEquationAst::Type::FALSE, astParent);
+    } else if (node->isMathmlElement("exponentiale")) {
+        ast->mPimpl->populate(AnalyserEquationAst::Type::E, astParent);
+    } else if (node->isMathmlElement("pi")) {
+        ast->mPimpl->populate(AnalyserEquationAst::Type::PI, astParent);
+    } else if (node->isMathmlElement("infinity")) {
+        ast->mPimpl->populate(AnalyserEquationAst::Type::INF, astParent);
+    } else if (node->isMathmlElement("notanumber")) {
+        ast->mPimpl->populate(AnalyserEquationAst::Type::NAN, astParent);
+    }
+}
+
+void Analyser::AnalyserImpl::processComponent(const ComponentPtr &component)
+{
+    // Retrieve the math string associated with the given component and process
+    // it, one equation at a time.
+
+    auto xmlDoc = std::make_shared<XmlDoc>();
+    auto math = component->math();
+
+    if (!math.empty()) {
+        xmlDoc->parseMathML(math, false);
```

Comment:
Does this need to iterate through the child math nodes, as there could be any number of them, or have they been merged into one by the time this is called?

Reply:
Argh, good point, they should indeed be merged into one. This is probably something that `XmlDoc::parseMathML()` should do in fact. Indeed, we use that method to validate a MathML string, but I am not sure it does a good of it if we have more than one `<math>` block...? @hsorby? @nickerso?

Reply:
Not sure. Ping @hsorby and @nickerso.

Reply:
It does need addressing.  Either with a test or with an issue.

Reply:
Ok, I am going to address it in this PR. Hopefully, this shouldn't be too much work... *Fingers crossed!*

Reply:
they shouldn't be merged into one in general, right? Here you should just be iterating over all `math` nodes, which may be the single root node or there could be a number of them. `parseMathML` certainly doesn't look like it is doing any merging...

Reply:
Agreed that they shouldn't be merged upon calling `Component::setMath()`. However, when it comes to the analysis, we should indeed merge them. (This will be internal to the analyser, so the merged math will never be exposed.)

Reply:
Ok, done in commit e923d0ab.